### ISSUE TITLE
Feature Request: Support -Skip and -Pending on Context and Describe

### ISSUE THREAD

**Issue by RemcoKapinga** [pester/Pester#856]
It would be nice if we can use -Skip or -Pending on groups of tests, without having to specify -Skip for every individual test.

We now have to specify -Skip for every 'It':

    Describe 'SkipEveryIt' {

        BeforeAll { 
            $session = Try { New-PsSession -ComputerName $server -ErrorAction SilentlyContinue } Catch { $null } 
            $skip = ($null -EQ $session)
        }
        AfterAll  { if($session) { Remove-PsSession -Session $session -ErrorAction SilentlyContinue } }

        Context 'Networking' {
            It 'something remote .. ' -Skip:$skip { }
            It 'some other thing remote .. ' -Skip:$skip { }
            It 'some more thing remote .. ' -Skip:$skip { }
        }
    }

It would be nicer if we could use:

    Describe 'SkipContext' {

        BeforeAll { 
            $session = Try { New-PsSession -ComputerName $server -ErrorAction SilentlyContinue } Catch { $null } 
            $skip = ($null -EQ $session)
        }
        AfterAll  { if($session) { Remove-PsSession -Session $session -ErrorAction SilentlyContinue } }

        Context 'Networking' -Skip:$skip {{
            It 'something remote .. '{ }
            It 'some other thing remote .. ' { }
            It 'some more thing remote .. ' { }
        }
    }

**Comment by nohwnd** [comment]
I think that is doable, but the setups will still be run, because we don't know how many context/describes are in the current block. Is that fine? To me that would be unexpected.

**Comment by Remco-74** [comment]
If I could use -Skip on a Context, i would expect the following effects:
- No code in the block would be executed (so no before/after all, not tests etc..).
- All tests in the block will be marked as skipped). 
- And this would work recursively (for nested blocks).

**Comment by nohwnd** [comment]
If all code was in dome kind of block Pester controls we could do this easily, but that is not how people currently write the code, and we cannot force it on them. We could make it possible if you follow rules, but there will be a big warning sign on it. See #721 for discussion about running single test (which would make implementing your proposal easy).

**Comment by DarkLite1** [comment]
I was looking to open a new issue for this, but luckily I'm not the first one requesting this new feature. It would be really nice if we can use `-Skip` on a `Describe` or `Context` clause.

Currently, we're developing code in `VS Code` or in the `PowerShell ISE` and simply press `F5` to run it. A small demo piece of code:

```
$Skip = $true

Describe 'printer ports' {
    Context 'a correct printer port is' {
        # Set the stage

        ."$here\$sut" @testParams

        it 'still installed and left untouched' {
            $testResult = Get-PrinterPort -Name $testPrinter.PortName
            $testResult.Name | Should -Be $testPrinter.PortName
            $testResult.PrinterHostAddress | Should -Be $testPrinter.PortHostAddress
        } -Skip:$Skip
        it 'not registered as an action' {
            $Printers.Action | Where-Object {$_ -like '*printer port*'} | Should -BeNullOrEmpty
        } -Skip:$Skip
    }
    Context 'a missing printer port is' {
        # Set the stage

        ."$here\$sut" @testParams

        it 'installed with the correct PostHostAddress' {
            $testResult = Get-PrinterPort -Name $testPrinter.PortName -EA Ignore
            $testResult.Name | Should -Be $testPrinter.PortName
            $testResult.PrinterHostAddress | Should -Be $testPrinter.PortHostAddress
        } -Skip:$Skip
        it 'registered as an action' {
            $Printers | Where-Object {$_.Action -contains 'Installed printer port'} | Should -Not -BeNullOrEmpty
        } #-Skip:$Skip
    }
}
```

The `$Skip` variable is very important as it allows us to test a single piece of code by pressing `F5` instead of having to run them all. It doesn't bother us that it says ',skipped` for all the other tests. In the example above we are working on the last piece, that's why the `-Skip` is commented out.

So it would be great if we could skip a whole `Describe` section as it takes most of the time to run. Thank you for having another look at this.

**Comment by nohwnd** [comment]
@DarkLite1 In v5 there is a feature that is better suited for this, the `-Focus` paremeter. Putting `-Focus` on a test (or block) will run just it. There is no interactive mode (pressing F5), yet. But it works across multiple files. :)

**Comment by DarkLite1** [comment]
So it means that in the future you just add the switch `-Focus` to a `Context`, `Describe` or `It` clause and it will just run that piece of code? That sounds promising!

Related question perhaps, in `VS Code` there is also an option to run a specific piece of code. However, for me it seems like it doesn't do anything when I press `Run tests`:
![image](https://user-images.githubusercontent.com/10828570/54277886-4c0cea00-4591-11e9-954e-97216180bfa9.png)

Results in:
![image](https://user-images.githubusercontent.com/10828570/54277935-68108b80-4591-11e9-85b5-d9cb812e4def.png)

**Comment by nohwnd** [comment]
Yup. That is how it works. :) In the VS code the plugin (code lens it's called imho) will need to add support for v5, because right now they are just running the whole file with "test name" filter.

**Comment by DarkLite1** [comment]
Made a comment [here](https://github.com/PowerShell/vscode-powershell/issues/1500) because I think something is not ok in `VS Code` for running isolated tests.

**Comment by DarkLite1** [comment]
I like the idea of `V5` (Pester managed code) so I'm now wrapping my independent code in a `BeforeAll` clause that is managed by Pester. So we're fully compatible with the newer release when it comes. Thank you for the [examples and explanation](https://github.com/pester/Pester/tree/v5.0) you posted, it clarifies some best practices for the future. 

So crossing my fingers that:
- Running interactive with `F5` to just run one test will become possible
- The `VS Code` guys can fix the clickable `Run tests` button

That would be a developers dream come true :) 

Thank you Jakub for the magnificent work you do on this module. Really love it! Also the idea with the `TestCases` on the `it` block will be so much easier without the `Param` clause.

**Comment by nohwnd** [comment]
@DarkLite1 you are welcome :)

**Comment by nohwnd** [comment]
There is -Skip on Describe and Context, I don't plan to add -Pending.
